File: core/src/main/java/pcb/uwu/amount/quantity/Scalar.java
package pcb.uwu.amount.quantity;

import pcb.uwu.amount.finance.InterestRate;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.finance.InterestRateUnit;
import pcb.uwu.units.quantity.ScalarUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.units.quantity.ScalarUnit.SCALAR;

public class Scalar extends CompositeUnitAmount<ScalarUnit> {

	// region constructors

	public Scalar(Number value) {
		super(value, SCALAR);
	}

	public Scalar(Number value, Magnitude magnitude) {
		super(value, magnitude, SCALAR);
	}

	public Scalar(String value) {
		super(value, SCALAR);
	}

	public Scalar(String value, Magnitude magnitude) {
		super(value, magnitude, SCALAR);
	}

	public Scalar(BigDecimal value) {
		super(value, SCALAR);
	}

	public Scalar(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, SCALAR);
	}

	public Scalar(BigDecimalAmount amount) {
		super(amount, SCALAR);
	}

	public Scalar(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, SCALAR);
	}


	// endregion

	// region implement UnitAmount

	@Override
	public Scalar plus(UnitAmount<ScalarUnit> other, MathContext mathContext) {
		return new Scalar(plusAmount(other, mathContext));
	}

	@Override
	public Scalar minus(UnitAmount<ScalarUnit> other, MathContext mathContext) {
		return new Scalar(minusAmount(other, mathContext));
	}

	@Override
	public Scalar multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Scalar(multipliedByScalar(other, mathContext));
	}

	@Override
	public Scalar dividedBy(BigDecimal other, MathContext mathContext) {
		return new Scalar(dividedByScalar(other, mathContext));
	}

	@Override
	public Scalar convertTo(ScalarUnit unit) {
		return new Scalar(getAmountIn(unit));
	}

	// endregion

	// region composition

	public InterestRate dividedBy(Time time, MathContext mathContext) {
		BigDecimalAmount amount = getAmount()
				.dividedBy(time.getAmount().getValue(), mathContext);
		InterestRateUnit unit = new InterestRateUnit(time.getUnit());

		return new InterestRate(amount, unit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/base/Scalar.kt
package pcb.uwu.amount.base

import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.ScalarUnit
import pcb.uwu.util.UnitAmountUtils

open class Scalar : CompositeUnitAmount<ScalarUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: ScalarUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: ScalarUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(scalar: UnitAmount<ScalarUnit>) =
        Scalar(amount = this.amount + (scalar to this.unit).amount,
               unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(scalar: UnitAmount<ScalarUnit>) =
        Scalar(amount = this.amount - (scalar to this.unit).amount,
               unit = this.unit)


    override fun times(scalar: Number) =
        Scalar(amount = this.amount * scalar,
               unit = this.unit)

    override fun div(scalar: Number) =
        Scalar(amount = this.amount / scalar,
               unit = this.unit)

    override fun to(unit: ScalarUnit) =
        Scalar(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
               unit = this.unit)

    // endregion
}
